File: src/server/internal/server/mapcolliders.go
package server

import (
	"encoding/json"

	"pixi_game_server/internal/collision"
)

type mapColliderView struct {
	ColliderID uint64 `json:"colliderId"`
	MinX       int32  `json:"minX"`
	MinY       int32  `json:"minY"`
	MaxX       int32  `json:"maxX"`
	MaxY       int32  `json:"maxY"`
	RenderKind string `json:"renderKind"`
}

type mapCollidersView struct {
	CampaignID       uint64            `json:"campaignId"`
	GeneratorVersion string            `json:"generatorVersion"`
	MapVersion       string            `json:"mapVersion"`
	CellSize         int32             `json:"cellSize"`
	PlayerRadius     int32             `json:"playerRadius"`
	Colliders        []mapColliderView `json:"colliders"`
}

// BuildMapCollidersJSON marshals the immutable /api/map-colliders payload
// described in docs/collisions_plan.md. colliders must already be validated
// and ordered by collider ID (the same order the server's MapStaticGrid was
// built from).
func BuildMapCollidersJSON(campaignID uint64, generatorVersion, mapVersion string, colliders []collision.MapAABB) ([]byte, error) {
	view := mapCollidersView{
		CampaignID:       campaignID,
		GeneratorVersion: generatorVersion,
		MapVersion:       mapVersion,
		CellSize:         collision.GridCellSize,
		PlayerRadius:     collision.PlayerRadius,
		Colliders:        make([]mapColliderView, len(colliders)),
	}
	for i, c := range colliders {
		view.Colliders[i] = mapColliderView{
			ColliderID: c.ID,
			MinX:       c.MinX,
			MinY:       c.MinY,
			MaxX:       c.MaxX,
			MaxY:       c.MaxY,
			RenderKind: c.RenderKind,
		}
	}
	return json.Marshal(view)
}

File: src/shared/mapColliders.ts
import { WORLD } from "./gameConfig";
import { showLoadingError } from "./loadingScreen";
import { MapStaticGrid, type MapAABB } from "../client/collision/grid";

interface MapCollidersResponse {
    campaignId: number;
    generatorVersion: string;
    mapVersion: string;
    cellSize: number;
    playerRadius: number;
    colliders: {
        colliderId: number;
        minX: number;
        minY: number;
        maxX: number;
        maxY: number;
        renderKind: string;
    }[];
}

async function loadMapColliders(): Promise<MapCollidersResponse> {
    try {
        const res = await fetch("/api/map-colliders");
        if (!res.ok) throw new Error(`status ${res.status}`);
        return (await res.json()) as MapCollidersResponse;
    } catch (err) {
        showLoadingError("Could not reach the game server. Please try reloading.");
        throw err;
    }
}

const mapCollidersResponse: MapCollidersResponse = await loadMapColliders();

export const CAMPAIGN_ID = mapCollidersResponse.campaignId;
export const MAP_VERSION = mapCollidersResponse.mapVersion;

const mapColliders: MapAABB[] = mapCollidersResponse.colliders.map((c) => ({
    id: c.colliderId,
    minX: c.minX,
    minY: c.minY,
    maxX: c.maxX,
    maxY: c.maxY,
    renderKind: c.renderKind,
}));

/**
 * The immutable, campaign-wide static map collider index. Built once at
 * startup from GET /api/map-colliders, mirroring the server's
 * MapStaticGrid — see docs/collisions_plan.md.
 */
export const mapStaticGrid = new MapStaticGrid(WORLD.virtualSize.width, WORLD.virtualSize.height, mapColliders);
